ArduCopter/Copter.cpp

```diff
@@ -20,6 +82,515 @@
+
+    // exit immediately if no new radio frame or not in simple mode
+    if (ap.simple_mode == 0 || !ap.new_radio_frame) {
+        return;
+    }
+
+    // mark radio frame as consumed
+    ap.new_radio_frame = false;
+
+    if (ap.simple_mode == 1) {
+        // rotate roll, pitch input by -initial simple heading (i.e. north facing)
+        rollx = channel_roll->get_control_in()*simple_cos_yaw - channel_pitch->get_control_in()*simple_sin_yaw;
+        pitchx = channel_roll->get_control_in()*simple_sin_yaw + channel_pitch->get_control_in()*simple_cos_yaw;
+    }else{
+        // rotate roll, pitch input by -super simple heading (reverse of heading to home)
+        rollx = channel_roll->get_control_in()*super_simple_cos_yaw - channel_pitch->get_control_in()*super_simple_sin_yaw;
+        pitchx = channel_roll->get_control_in()*super_simple_sin_yaw + channel_pitch->get_control_in()*super_simple_cos_yaw;
+    }
+
+    // rotate roll, pitch input from north facing to vehicle's perspective
+    channel_roll->set_control_in(rollx*ahrs.cos_yaw() + pitchx*ahrs.sin_yaw());
+    channel_pitch->set_control_in(-rollx*ahrs.sin_yaw() + pitchx*ahrs.cos_yaw());
+}
+
+// update_super_simple_bearing - adjusts simple bearing based on location
+// should be called after home_bearing has been updated
+void Copter::update_super_simple_bearing(bool force_update)
+{
+    if (!force_update) {
+        if (ap.simple_mode != 2) {
+            return;
+        }
+        if (home_distance() < SUPER_SIMPLE_RADIUS) {
+            return;
+        }
+    }
+
+    const int32_t bearing = home_bearing();
+
+    // check the bearing to home has changed by at least 5 degrees
+    if (labs(super_simple_last_bearing - bearing) < 500) {
+        return;
+    }
+
+    super_simple_last_bearing = bearing;
+    const float angle_rad = radians((super_simple_last_bearing+18000)/100);
+    super_simple_cos_yaw = cosf(angle_rad);
+    super_simple_sin_yaw = sinf(angle_rad);
+}
+
+void Copter::read_AHRS(void)
+{
+    // Perform IMU calculations and get attitude info
+    //-----------------------------------------------
+#if HIL_MODE != HIL_MODE_DISABLED
+    // update hil before ahrs update
+    gcs().update();
+#endif
+
+    // we tell AHRS to skip INS update as we have already done it in fast_loop()
+    ahrs.update(true);
+}
+
+// read baro and log control tuning
+void Copter::update_altitude()
+{
+    // read in baro altitude
+    read_barometer();
+
+    if (should_log(MASK_LOG_CTUN)) {
+        Log_Write_Control_Tuning();
+    }
+}
+
+#if OSD_ENABLED == ENABLED
+void Copter::publish_osd_info()
+{
+    AP_OSD::NavInfo nav_info;
+    nav_info.wp_distance = flightmode->wp_distance() * 1.0e-2f;
+    nav_info.wp_bearing = flightmode->wp_bearing();
+    nav_info.wp_xtrack_error = flightmode->crosstrack_error() * 1.0e-2f;
+    nav_info.wp_number = mode_auto.mission.get_current_nav_index();
+    osd.set_nav_info(nav_info);
+}
+#endif
+
 /*
   constructor for main Copter class
  */
```

Comment:
Nitpicking: I would put the constructor as the first thing on the file.

Reply:
I put it down where we created the instance of it on purpose.

@rmackay9 - preferences?